### PR TITLE
configurable publication repository

### PR DESCRIPTION
**Pull request by StefMa**
Rest of the implementation of #430 :)

### CODE REVIEW

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java [hunk 1]
```diff
@@ -61,7 +64,12 @@ public void execute(JavaBintrayPlugin plugin) {
                         deferredConfiguration(subproject, new Runnable() {
                             public void run() {
                                 UpdateReleaseNotesTask updateNotes = (UpdateReleaseNotesTask) project.getTasks().getByName(ReleaseNotesPlugin.UPDATE_NOTES_TASK);
-                                updateNotes.setPublicationRepository(BintrayUtil.getRepoLink(bintray));
+                                String userSpecifiedRepo = conf.getLenient().getReleaseNotes().getPublicationRepository();
+                                if (userSpecifiedRepo != null) {
```
Comment (wwilk) [inline subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: Why lenient config here, while regular one in GradlePortalReleasePlugin?

Comment (StefMa) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: Sry. I don't really get it 🙃 
That was original designed by @szczepiq.

Should I change that code to
```
(UpdateReleaseNotesTask) project.getTasks().getByName(ReleaseNotesPlugin.UPDATE_NOTES_TASK);
```
as well? 🤔

Comment (wwilk) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: It would be also good to create unit tests for these two cases.

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePlugin.java [hunk 2]
```diff
@@ -59,6 +64,9 @@ public void execute(Plugin plugin) {
                         Task archivesTask = subproject.getTasks().getByName("buildArchives");
                         publishPlugins.dependsOn(archivesTask);
                         gitPush.mustRunAfter(archivesTask);
+
+                        UpdateReleaseNotesTask updateNotes = (UpdateReleaseNotesTask) project.getTasks().getByName(ReleaseNotesPlugin.UPDATE_NOTES_TASK);
+                        updateNotes.setPublicationRepository(conf.getReleaseNotes().getPublicationRepository());
```
Comment (wwilk) [inline subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePlugin.java]: Also missing unit test here

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java (hunk 1 above)
Comment (StefMa) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: Hmm.. wait 🤔 
We have to get the `publicationRepository` from the config here.
Because we want to set it to the task :)
We can't use the task to get the `publicationRepository`. It is always `null` ...

Comment (wwilk) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: I'm only asking about lenient config here: 
`conf.getLenient().getReleaseNotes().getPublicationRepository()`
Why here it's lenient, and in GradlePortalReleasePlugin it's not lenient. 
In general we use lenient one when the property can be null. But you are right, it was like that in the first commit for this issue.
@szczepiq Why is it different for these two classes?

Comment (szczepiq) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: Sure!

lenient() is documented in the Javadoc, let me know if it is unclear (or suggest an improvement / rename).

In this case, we should be using lenient() because repo is optional - we can get the value automatically from bintray extension. In the other case (gradle portal plugin), it is not optional, we cannot get the value from bintray extension because bintray plugin is not used.

Does it sound right? Good question!

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java [hunk 3]
```diff
@@ -61,7 +64,12 @@ public void execute(JavaBintrayPlugin plugin) {
...
+                                    updateNotes.setPublicationRepository(userSpecifiedRepo);
+                                } else {
```
Comment (szczepiq) [inline subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: Unit test please :) Let us know if you need help.

File: subprojects/shipkit/src/main/groovy/org/shipkit/gradle/configuration/ShipkitConfiguration.java [hunk 4]
```diff
@@ -255,6 +255,24 @@ public void setLabelMapping(Map<String, String> labelMapping) {
         public void setIgnoreCommitsContaining(Collection<String> commitMessageParts) {
             store.put("releaseNotes.ignoreCommitsContaining", commitMessageParts);
         }
+
+        /**
+         * Set the Publication Repository where your project can be found
```
Comment (szczepiq) [inline subprojects/shipkit/src/main/groovy/org/shipkit/gradle/configuration/ShipkitConfiguration.java]: The publication repository is the place where to look for binaries. It could be MavenCentral, Bintray or some other repository. Can you update the javadoc accordingly? Can you add an example in the Javadoc? See an example if Shipkit project itself.

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePlugin.java (hunk 2 above)
Comment (szczepiq) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/release/GradlePortalReleasePlugin.java]: +1

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java (hunk 3 above)
Comment (StefMa) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePlugin.java]: A little bit help would be great 👍 
I've never used spock before. I've also never tested a gradle plugin 🙃 
I've tried it with something like that:
```
    def "applies"() {
        expect:
        project.plugins.apply(BintrayReleasePlugin)
        project.plugins.apply(JavaBintrayPlugin)
        //def repository = project.tasks.getByName("updateReleaseNotes").getPublicationRepository()
        //assert repository == null
        project.evaluate()
        def repositoryNonNull = project.tasks.getByName("updateReleaseNotes").getPublicationRepository()
        assert repositoryNonNull != null
    }
```
But unfortunately that isn't working..

**Comment by StefMa** [comment]
Wow. Thanks for the help @szczepiq 💪 😗 
I've cherry-picked your example test and adapt it to the new needs.
Hope that don't lead into merge conflicts if #456 got merged 🙂

Hope everything is fine now 😇

File: subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePluginTest.groovy [hunk 5]
```diff
@@ -1,17 +1,45 @@
...
+        conf.releaseNotes.publicationRepository = "publicRepo"
 
-    def "applies"() {
-        expect:
         project.plugins.apply(BintrayReleasePlugin)
         project.plugins.apply(JavaBintrayPlugin)
+
+        project.bintray.pkg.userOrg = "some-org"
+        project.bintray.pkg.repo = "some-repo"
+        project.bintray.pkg.name = "some-pkg"
```
Comment (wwilk) [inline subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/bintray/BintrayReleasePluginTest.groovy]: Configuration of bintray extension seems irrelevant for this test. UpdateReleaseNotesTask will use conf.releaseNotes.publicationRepository to set the value of publicationRepository.

**Comment by StefMa** [comment]
Because of merge conflicts I've rebased the branch.
I've also removed the configuration of the bintray plugin in the test (see https://github.com/mockito/shipkit/pull/446#discussion_r142036445)

**Comment by wwilk** [comment]
Thanks! Merging

**Comment by szczepiq** [comment]
@StefMa, Can you fix Shipkit so that it can consume new version? After this PR was merged, Shipkit no longer can consume Shipkit :) See the build failure: https://travis-ci.org/mockito/shipkit/builds/282839303?utm_source=github_status&utm_medium=notification
